### PR TITLE
Feature: Add colorization to js doc comments

### LINKED ISSUE #1347: Better colorization of doc comments
We just color them as regular comments now. We don't, for example, change the color of doc tags.

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix #1347

Add syntax highlighting to doc comments. Include the syntax definition for the tmlanguage as well as the semantic highlighting.

image

[Try it in the playground](https://cadlplayground.z22.web.core.windows.net/prs/2306/?c=LyoqCiAqIEB0ZW1wbGF0ZSBUIFRob3NlIGFyZSBrbm93biB0YWdzIHNvIHdlIGhpZ2hsaWdodCBsaWtlIGtleXdvcmTFQXBhcmFtIGZvbyBUaGlzIGlzIGHEDsscYmFyIC3LHmJhcgogxn5yZXR1cm5zIMgIc3R1ZmbIHWV4YW1wbGXJOm5vdCBhIGRlZmluZWTkAJ7xAJ1kaWZmZXJlbnRseQogKi8KbW9kZWzEcCB7fQo%3D&e=%40typespec%2Fopenapi3&options=%7B%7D)

### CODE REVIEW

File: packages/compiler/src/server/tmlanguage.ts [hunk 1]
```diff
@@ -160,12 +168,62 @@ const blockComment: BeginEndRule = {
   end: "\\*/",
 };
 
+const docCommentParam: MatchRule = {
```
Comment (markcowl) [inline packages/compiler/src/server/tmlanguage.ts:171]: Seems like we should have at least one test for an unknown tag mapping to the new `entity.name.function.macro.tsp`

Comment (timotheeguerin) [reply packages/compiler/src/server/tmlanguage.ts:171]: it does have it with `@custom`

**Comment by bterlson** [comment]
Sorry I missed this until now.

I love the colors, definitely. However, I am pretty concerned about coloring the jsdoc keys the same as keywords. In the context of the given example, `model` gets lost. But I think the problem is worse in model properties where the jsdoc keywords will be much more prominent than the property itself.

Is there another option that might make sense?

**Comment by bterlson** [comment]
Eh, looking at it again it looks ok for the model case. Maybe my concern is really just that member names aren't rendered differently and get lost in the list of comments and decorators.

**Comment by timotheeguerin** [comment]
> Sorry I missed this until now.
> 
> I love the colors, definitely. However, I am pretty concerned about coloring the jsdoc keys the same as keywords. In the context of the given example, `model` gets lost. But I think the problem is worse in model properties where the jsdoc keywords will be much more prominent than the property itself.
> 
> Is there another option that might make sense?

there is definitely some other token we can use than keyword I was just trying to keep the same used in typescript 
image

we could not special case the known ones so they all like th `@example` in the image above

**Comment by allenjzhang** [comment]
like the feature but not sure about the color choices. One of the key benefit for dev doc is to have comment color to allow eyes to easily differentiate between comments and code. So now we want to differentiate tokens in comments further, I would like to suggest color from a muted palette. Maybe from gray or darker (than comment green) green?

**Comment by timotheeguerin** [comment]
> like the feature but not sure about the color choices. One of the key benefit for dev doc is to have comment color to allow eyes to easily differentiate between comments and code. So now we want to differentiate tokens in comments further, I would like to suggest color from a muted palette. Maybe from gray or darker (than comment green) green?

We don't really control the actual color. The only thing we can do is give a token for what that chunk of text is. It is then up to the theme to color it. Some theme support more or less coloring. I did give it a different token than keywords but it default to that kind. This is also the same tokenization as typescript does with JSDoc. We could default to another token than keyword but it would also either could be a incorrect meaning and result in completely different colors between themes.

**Comment by allenjzhang** [comment]
> > like the feature but not sure about the color choices. One of the key benefit for dev doc is to have comment color to allow eyes to easily differentiate between comments and code. So now we want to differentiate tokens in comments further, I would like to suggest color from a muted palette. Maybe from gray or darker (than comment green) green?
> 
> We don't really control the actual color. The only thing we can do is give a token for what that chunk of text is. It is then up to the theme to color it. Some theme support more or less coloring. I did give it a different token than keywords but it default to that kind. This is also the same tokenization as typescript does with JSDoc. We could default to another token than keyword but it would also either could be a incorrect meaning and result in completely different colors between themes.

**Review by markcowl — APPROVED** [review]

**Pull request merged**